File: Mar-18/19-03-18/java/01-Intro-to-Dynamic-programming-1.java
import java.util.*;

class TestClass {
    
    public static int MOD = 1000000007;
    
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner( System.in );
        TestClass obj = new TestClass();
        int n = input.nextInt();
        int ar[] = new int[n];
        int maxi = 0;
        for(int i = 0; i < n; i++)
        {
            ar[i] = input.nextInt();
            if(ar[i] > maxi)
                maxi = ar[i];
        }
        long dp[] = new long[maxi+1];
        dp[0] = 1;
        for(int i = 1; i <= maxi; i++)
        {
            dp[i] = ( (dp[i-1] % MOD) * (i % MOD) ) % MOD;
        }
        
        
        for(int i = 0; i < n; i++)
        {
            System.out.println(dp[ ar[i] ]);
        }
        
    }
}

File: Mar-18/19-03-18/python/01-Intro-to-Dynamic-programming-1.py
MOD = 1000000007
dp = {}
dp[0] = dp[1] = 1

def solu(n):
    for i in range(2,n+1):
        dp[i] = ( (dp[i-1] % MOD) * (i % MOD) ) % MOD

n = int(input())
ar = []

for i in range(0,n):
    ar.append( int(input()) )

solu(max(ar))

for i in range(0,n):
    print(dp[ ar[i] ])
